Detectors/CTF/workflow/src/CTFWriterSpec.cxx

```diff
@@ -486,7 +525,63 @@ void CTFWriterSpec::closeDictionaryTreeAndFile(CTFHeader& header)
 }
 
 //___________________________________________________________________
-DataProcessorSpec getCTFWriterSpec(DetID::mask_t dets, uint64_t run, bool doCTF, bool doDict, bool dictPerDet, size_t szmn, size_t szmx)
+void CTFWriterSpec::createLockFile(const o2::header::DataHeader* dh, int level)
+{
+  // create lock file for the CTF to be written to the storage of given level
+  mLockFileName = fmt::format("{}/ctfs{}-{}_{}_{}_{}.lock", mLockFileDir, level, o2::utils::Str::getRandomString(8), mRun, dh->firstTForbit, dh->tfCounter);
+  mLockFD = open(mLockFileName.c_str(), O_RDWR | O_CREAT | O_CLOEXEC, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);
+  if (mLockFD == -1) {
+    throw std::runtime_error(fmt::format("Error opening lock file {}", mLockFileName));
+  }
+  if (lockf(mLockFD, F_LOCK, 0)) {
+    throw std::runtime_error(fmt::format("Error locking file {}", mLockFileName));
+  }
+}
+
+//___________________________________________________________________
+void CTFWriterSpec::removeLockFile()
+{
+  // remove CTF lock file
+  if (mLockFD != -1) {
+    if (lockf(mLockFD, F_ULOCK, 0)) {
+      throw std::runtime_error(fmt::format("Error unlocking file {}", mLockFileName));
+    }
+    mLockFD = -1;
+    std::filesystem::remove(mLockFileName);
+  }
+}
+
+//___________________________________________________________________
+size_t CTFWriterSpec::getAvailableDiskSpace(const std::string& path, int level)
+{
+  // count number of CTF files in processing (written to storage at given level) from their lock files
+  std::regex pat{fmt::format("({}/ctfs{}-[[:alnum:]_]+\\.lock$)", mLockFileDir, level)};
+  int nLocked = 0;
+  size_t written = 0;
+  for (const auto& entry : std::filesystem::directory_iterator(mLockFileDir)) {
+    const auto& entryName = entry.path().native();
+    if (std::regex_search(entryName, pat) && (mLockFD < 0 || entryName != mLockFileName)) {
+      int fdt = open(entryName.c_str(), O_RDONLY);
+      if (lockf(fdt, F_TEST, 0) != 0) {
+        nLocked++;
+        size_t sz = 0;
+        auto nrd = read(fdt, &sz, sizeof(size_t));
+        if (nrd == sizeof(size_t)) {
+          written += sz;
+        }
+      }
+      close(fdt);
+    }
+  }
+  std::error_code ec;
+  const auto si = std::filesystem::space(path, ec);
+  int64_t avail = int64_t(si.available) - nLocked * mChkSize + written; // account already written part of unfinished files
```

Comment:
In this marginal case indeed there would be a double counting and the writer may switch to real disk. Just counting number of files would not be reliable: one would need to count (1) finished but not yet evacuated files, (2) files currently in writing (with locks)  + wathever other files we may decide to have on the RAM disk or garbage (e.g. ctf*.part files from crashed jobs). And this will become even more complicated if there are directories on the RAM disk.